rundeck_action/action_types/rundeck.ts

```diff
@@ -61,164 +53,215 @@ export function getActionType({ logger }: { logger: Logger }): ActionType {
+  }
 
-    if (error.response) {
-      const { status, statusText, headers: responseHeaders } = error.response;
-      const message = `[${status}] ${statusText}`;
+  // incident list is empty
+  if (!pdIncidentListResult.data.incidents.length) {
+    const message: string = `Pager Duty incident list requested by dedupKey(incident_key), "${dedupKey}", is empty.`
+    logger.warn(`error on ${actionId} rundeck action: ${message}`);
+    return errorPagerDutyProcess(actionId, message);
+  }
 
-      logger.warn(`error on ${actionId} rundeck event: ${message}`);
+  // get incident id
+  const incidentId = pdIncidentListResult.data.incidents.pop().id
 
-      // The request was made and the server responded with a status code
-      // that falls out of the range of 2xx
-      // special handling for 5xx
-      if (status >= 500) {
-        return retryResult(actionId, message);
-      }
+  // call Pager Duty add note API
+  try {
+    const pagerDutyApiOptions = {
+      headers,
+      pdApiKey,
+      incidentId,
+      executionLink,
+    };
 
-      return errorResultInvalid(actionId, message);
-    }
+    await addNoteToPagerDutyIncident(pagerDutyApiOptions);
 
-    const message = 'Unreachable rundeck host, are you sure the address is correct?'
-    
-    logger.warn(`error on ${actionId} rundeck action: ${message}`);
+    logger.info(`Calling pagerduty "create a note API" step succeeded in rundeck action "${actionId}"`);
+
+  } catch (err) {
 
-    return errorResultUnreachable(actionId, message);
+    const message = err.response ? `${err.response.status} ${err.response.data.error.message}` : err.message;
+    logger.warn(`error on ${actionId} rundeck action: an error occurred while calling pager duty API: ${message}`);
+    return errorPagerDutyProcess(actionId, err.message);
   }
+
+  logger.info(`response from rundeck action "${actionId}": [HTTP ${rundeckResult.status}] ${rundeckResult.statusText}`);    
+  
+  return successResult(rundeckResult.data);
+  
+}
+
+async function executeRundeckJob(rundeckApiOptions: any): Promise<AxiosResponse|AxiosError> {
+
+  const {
+    headers, 
+    rundeckBaseUrl,
+    rundeckApiToken, 
+    rundeckApiVersion, 
+    rundeckJobId, 
+    jobParams,
+  } = rundeckApiOptions;
+
+  // add rundeck api token to headers
+  Object.assign(headers, {"X-Rundeck-Auth-Token": rundeckApiToken});
+
+  // trim last '/' in the rundeckBaseUrl
+  const rundeckBaseUrlWithoutSlash = rundeckBaseUrl.endsWith('/') ? rundeckBaseUrl.slice(0, -1) : rundeckBaseUrl;
+
+  const rundeckApiUrl = `${rundeckBaseUrlWithoutSlash}/api/${rundeckApiVersion}/job/${rundeckJobId}/executions`;
+
+  return axios.post(rundeckApiUrl, jobParams, { headers });
+}
+
+async function getPagerDutyIncidentList(pagerDutyApiOptions: any): Promise<AxiosResponse|AxiosError> {
+
+  const {headers, pdApiKey, dedupKey} = pagerDutyApiOptions;
+
+  Object.assign(headers, {authorization: `Token token=${pdApiKey}`});
```

Comment:
We're sharing a reference to a single `headers` object throughout this module. We're open to the potential of polluting this object across different methods. 

A good practice here would be to copy the `headers` locally when mutating.

```suggestion
  headers = Object.assign({}, headers, {authorization: `Token token=${pdApiKey}`});
```

Reply:
This is absolutely sweet! Thanks!